File: solutions/17/main.cpp
/* ************************************************************************
> File Name:     main.cpp
> Author:        Sean
> Project:    LeetCode
> Created Time:  Mon Nov  1 09:18:33 2021
> Description:   
 ************************************************************************/
#include <iostream>
#include <string> 
#include <vector> 
#include <gtest/gtest.h>

using namespace std; 

class Solution {
private:
    const string letterMap[10] = {
        "", // 0
        "", // 1
        "abc", // 2
        "def", // 3
        "ghi", // 4
        "jkl", // 5
        "mno", // 6
        "pqrs", // 7
        "tuv", // 8
        "wxyz", // 9
    };
public:
    vector<string> result;
    string path;
    void backtracking(const string& digits, int startId) {
        if (digits.size() == startId) {
            result.push_back(path);
            return;
        }
        int digit = digits[startId] - '0';
        string letters = letterMap[digit];
        for (int i = 0; i < letters.size(); ++i) {
            path.push_back(letters[i]);
            backtracking(digits, startId+1);
            path.pop_back();
        }
    }
    vector<string> letterCombinations(string digits) {
        if (digits.size() == 0) return result;
        backtracking(digits, 0);
        return result;
    }
};

TEST(Test17, SimpleTest) {
    const string digits = "23";
    Solution s; 
    vector<string> result = s.letterCombinations(digits);
    for (const auto& r : result) {
        cout << "comb ";
        for (const auto& a : r)
            cout << a;
        cout << endl; 
    }
}

int main() {
    testing::InitGoogleTest();
    return RUN_ALL_TESTS();
}

File: solutions/17/main.py
class Solution:
    def __init__(self): 
        self.path = ""
        self.result = []
        
        self.letterMap = {
            "2" : 'abc', 
            "3" : 'def', 
            "4" : 'ghi', 
            "5" : 'jkl', 
            "6" : 'mno', 
            "7" : 'pqrs', 
            "8" : 'tuv', 
            "9" : 'wxyz'
        }
    
    def backtracking(self, digits, startId): 
        if (startId == len(digits)):
            self.result.append(self.path)
            return 
        digit = digits[startId]
        letters = self.letterMap[digit] 
        for i in range(0, len(letters)):
            self.path += letters[i]
            self.backtracking(digits, startId+1)
            self.path = self.path[:-1]
        
    def letterCombinations(self, digits):
        if (len(digits) == 0): 
            return self.result
        self.backtracking(digits, 0) 
        return self.result

if __name__ == "__main__":
    s = Solution()
    print(s.letterCombinations("23"))
